language/stdlib/modules/validator_set.mvir

```diff
@@ -1,30 +1,101 @@
+        keys: Self.ValidatorPublicKeys,
+    }
+
     resource T {
-        validators: Vector.T<Self.TrustedPeer>,
+        validators: Vector.T<Self.ValidatorInfo>,
     }
 
-    make_new_validator_key(
-        account_address: address,
-        consensus_pubkey: bytearray,
+    // This can only be invoked by the special validator set address, and only a single time.
+    // Currently, it is invoked in the genesis transaction.
+    public initialize() {
+        // Only callable by the validator set address
+        assert(get_txn_sender() == 0x1, 1);
+
+        move_to_sender<T>(T { validators: Vector.empty<Self.ValidatorInfo>() });
+
+        return;
+    }
+
+    // Register the transaction sender as a candidate validator by creating a ValidatorConfig
+    // resource under their account
+    public register_candidate_validator(
         network_signing_pubkey: bytearray,
-        network_identity_pubkey: bytearray
-    ): Self.TrustedPeer {
-        let key: Self.TrustedPeer;
-        key = TrustedPeer {
-            account_address: move(account_address),
-            network_signing_pubkey: move(network_signing_pubkey),
-            network_identity_pubkey: move(network_identity_pubkey),
-            consensus_pubkey: move(consensus_pubkey),
-        };
-        return move(key);
+        network_identity_pubkey: bytearray,
+        consensus_pubkey: bytearray) {
+
+        move_to_sender<ValidatorConfig>(
+            ValidatorConfig {
+                keys: ValidatorPublicKeys {
+                    network_signing_pubkey: move(network_signing_pubkey),
+                    network_identity_pubkey: move(network_identity_pubkey),
+                    consensus_pubkey: move(consensus_pubkey),
+                }
+            });
+
+        return;
     }
+
+    // Get the consensus public key for the validator candidate at the given address
+    public get_consensus_pubkey(addr: address): bytearray acquires ValidatorConfig {
+        let config_ref: &Self.ValidatorConfig;
+        let keys_ref: &Self.ValidatorPublicKeys;
+
+        config_ref = borrow_global<ValidatorConfig>(move(addr));
+        keys_ref = &move(config_ref).keys;
+        return *&move(keys_ref).consensus_pubkey;
+    }
+
+    // Rotate a validator candidate's consensus public key. The change will not take effect until
+    // the next reconfiguration.
+    public rotate_consensus_pubkey(consensus_pubkey: bytearray) acquires ValidatorConfig {
+        let config_ref: &mut Self.ValidatorConfig;
+        let keys_ref: &mut Self.ValidatorPublicKeys;
+
+        config_ref = borrow_global_mut<ValidatorConfig>(get_txn_sender());
+        keys_ref = &mut move(config_ref).keys;
+        *(&mut move(keys_ref).consensus_pubkey) = move(consensus_pubkey);
+
+        return;
+    }
+
+    // TODO: Decide on access control policy. For now, we ensure that this is only callable from the
+    // genesis txn. Obviously, we'll need a different policy once we support reconfiguration.
```

Comment:
I remember one of our ideas were that a caller could that scans account addresses for LITs and ValidatorConfig and then just set the validators with that config if it changed?  I'm not sure how this method fits in.

Reply:
Good question. This fits into that model. Basically, the Association (or whoever has the authority to set the validator set) would write a transaction script with a list of a validator addresses and call `add_validator` for each address in the list. This `add_validator` code is what checks the LITs and ValidatorConfig.